pkg/collector/autodiscovery/autoconfig.go

```diff
@@ -209,32 +205,65 @@ func (ac *AutoConfig) getAllConfigs() []check.Config {
 				errorStats.removeConfigError(cfg.Name)
 			}
 		}
+		if withResolve {
+			resolvedConfigs := []check.Config{}
+			for _, config := range cfgs {
+				rc, err := ac.resolve(config)
+				if err != nil {
+					log.Error(err)
+				}
+				resolvedConfigs = append(resolvedConfigs, rc...)
+			}
 
-		configs = append(configs, cfgs...)
+			configs = append(configs, resolvedConfigs...)
+		} else {
+			configs = append(configs, cfgs...)
+		}
 	}
 
 	return configs
 }
 
-// getAllChecks gets all the check instances for any configurations found.
-func (ac *AutoConfig) getAllChecks() []check.Check {
-	all := []check.Check{}
-	configs := ac.getAllConfigs()
+// getChecksFromConfigs gets all the check instances for given configurations
+// optionally can populate ac cache config2checks
+func (ac *AutoConfig) getChecksFromConfigs(configs []check.Config, populateCache bool) []check.Check {
+	allChecks := []check.Check{}
 	for _, config := range configs {
+		configDigest := config.Digest()
 		checks, err := ac.GetChecks(config)
 		if err != nil {
 			continue
 		}
-
-		all = append(all, checks...)
+		for _, check := range checks {
+			allChecks = append(allChecks, check)
+			if populateCache {
+				// store the checks we schedule for this config locally
+				ac.config2checks[configDigest] = append(ac.config2checks[configDigest], check.ID())
+			}
+		}
 	}
 
-	return all
+	return allChecks
 }
 
-// resolveAndRun loads and resolves a given config and schedules the
-// corresponding Check instances.
-func (ac *AutoConfig) resolveAndRun(config check.Config) error {
+// schedule takes a slice of checks and schedule them
+func (ac *AutoConfig) schedule(checks []check.Check) {
+	for _, check := range checks {
+		log.Infof("Scheduling check %s", check)
+		_, err := ac.collector.RunCheck(check)
+		if err != nil {
+			log.Errorf("Unable to run Check %s: %v", check, err)
+			errorStats.setRunError(check.ID(), err.Error())
+			continue
+		}
+	}
+}
+
+// resolve loads and resolves a given config and can optionnaly schedules the
+// corresponding Check instances. Returns a slice of resolved configs
+func (ac *AutoConfig) resolve(config check.Config) ([]check.Config, error) {
+	configs := []check.Config{}
+
 	// add default metrics to collect to JMX checks
 	if config.CollectDefaultMetrics() {
```

Comment:
does it still make sense for this block to be here? What happens to JMX default metrics if `withResolve` was not passed?

Reply:
Not sure how this part works, should we move the `config.AddMetrics` part directly in `getAllConfigs`?
But I guess if it was not done in the first place there was a reason to store them through `name2jmxmetrics` between the two functions? 🤔

Reply:
@ISauve can you help?

Reply:
name2jmxmetrics exists because the file provider isn't responsible for resolving JMX conf files with their metrics files, so the idea was that when the file provider returns all the configs it finds (without knowing what it found) autodiscovery finds and saves all the jmx metric files in a map, and then when it's time to run the checks it will can quickly check if there are metrics for that check and add them if so

Reply:
If you're changing the flow such that this function is no longer run on all configs, then this should definitely be moved

Reply:
I guess we can move this to the `withResolve` part of `getAllConfigs`?

Reply:
You could add it `getAllConfigs`, but keep in mind that the configuration file which the metric file would need to be added to may not come from the same provider as the metric itself, so you can't really place this within the `for _, pd := range ac.providers { [collect] }` loop because you need both to be collected before checking if the metric file exists

Reply:
Do you ever call `getAllConfigs(false)`? I can only find instances where the parameter is `true`

Reply:
Nope for now it's only `getAllConfigs(true)` I thought it could be useful to keep a little separation with the `resolve` part but maybe there's no use case and we should always resolve the configs.

@masci WDYT?

Reply:
resolve is a no-op for non-template configs, it's fine to call it every time I think.

Reply:
Okay... so then there's 2 options:
1) take the `resolve` logic back out of `getAllConfigs` and add it after (like the old flow - calling `getAllConfigs` and then iterating over all the configs and resolving/running them), and keeping the JMX stuff in the `resolve` function with the AD stuff 
2) pulling the JMX-related block out of `resolve` and adding a separate function after `getAllConfigs` that exclusively attempts to resolve the JMX configs... but if we have to iterate over all the configs anyways, I think it makes more sense to just resolve the JMX and AD configs at the same time

Reply:
The key point is just that although the AD configs can be resolved whenever, the JMX ones require all configs to have been collected first, so the logic has to account for that

Reply:
@ISauve we can keep the `resolve` in `getAllConfigs` but do another loop after the loop on all providers if we want to keep the JMX code as is right?